### ISSUE TITLE
Strange error with numpy arrays and tuples

### ISSUE THREAD

**Issue by uweschmitt** [uqfoundation/dill#163]
The following code breaks with `dill 2.4` and `dill.2.5.0`:

```
import dill
import numpy as np

data = np.arange(1.0)
dill.dumps(data)

data = (np.arange(1.0),)
dill.loads(dill.dumps(data))
```

The stack trace is

```
Traceback (most recent call last):
  File "t.py", line 14, in <module>
    dill.loads(dill.dumps(data))
  File "/Users/uweschmitt/Projects/emzed2/lib/python2.7/site-packages/dill/dill.py", line 260, in loads
    return load(file)
  File "/Users/uweschmitt/Projects/emzed2/lib/python2.7/site-packages/dill/dill.py", line 250, in load
    obj = pik.load()
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 858, in load
    dispatch[key](self)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 1003, in load_tuple1
    self.stack[-1] = (self.stack[-1],)
IndexError: list index out of range
```

The error appears only in this combination. Both blocks run fine if executed alone.
The problem is that the second `dill.dumps(data)` creates invalid code `\x80\x02\x85q\x00.` (create tuple from last items on stack, but the stack is empty). I assume that during pickling the binary data from the array is not put onto the stack.

**Comment by uweschmitt** [comment]
For `protocol = 1` I get another error related to memoization:

````
Traceback (most recent call last):
  File "t.py", line 16, in <module>
    dill.loads(dill.dumps(data, protocol=1))
  File "/Users/uweschmitt/Projects/emzed2/lib/python2.7/site-packages/dill/dill.py", line 243, in dumps
    dump(obj, file, protocol, byref, fmode, recurse)#, strictio)
  File "/Users/uweschmitt/Projects/emzed2/lib/python2.7/site-packages/dill/dill.py", line 236, in dump
    pik.dump(obj)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 224, in dump
    self.save(obj)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 562, in save_tuple
    save(element)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/uweschmitt/Projects/emzed2/lib/python2.7/site-packages/dill/dill.py", line 232, in save_numpy_array
    pik.save_reduce(_create_array, (f, args, state, npdict), obj=obj)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 405, in save_reduce
    self.memoize(obj)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 244, in memoize
    assert id(obj) not in self.memo
AssertionError
````

**Comment by mmckerns** [comment]
Sorry for the long delay…

I can replicate this bug with:
```
>>> import dill
>>> import numpy as np
>>> x = np.array([1])
>>> y = (x,)
>>> dill.dumps(x)
'\x80\x02cdill.dill\n_create_array\nq\x00(cnumpy.core.multiarray\n_reconstruct\nq\x01cnumpy\nndarray\nq\x02K\x00\x85q\x03U\x01bq\x04\x87q\x05(K\x01K\x01\x85q\x06cnumpy\ndtype\nq\x07U\x02i8q\x08K\x00K\x01\x87q\tRq\n(K\x03U\x01<q\x0bNNNJ\xff\xff\xff\xffJ\xff\xff\xff\xffK\x00tq\x0cb\x89U\x08\x01\x00\x00\x00\x00\x00\x00\x00q\rtq\x0eNtq\x0fRq\x10.'
>>> dill.dumps(y)
'\x80\x02\x85q\x00.'
>>> dill.loads(_)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.6.dev0-py2.7.egg/dill/dill.py", line 273, in loads
    return load(file)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.6.dev0-py2.7.egg/dill/dill.py", line 263, in load
    obj = pik.load()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 864, in load
    dispatch[key](self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 1009, in load_tuple1
    self.stack[-1] = (self.stack[-1],)
IndexError: list index out of range
>>> 
```

A workaround seems to be possible:
```
>>> import StringIO as io
>>> f = io.StringIO()
>>> import dill
>>> import numpy as np
>>> x = np.array([1])
>>> y = (x,)
>>> p = dill.Pickler(f)
>>> p.dump(x)
>>> f.getvalue()
"cnumpy.core.multiarray\n_reconstruct\np0\n(cnumpy\nndarray\np1\n(I0\ntp2\nS'b'\np3\ntp4\nRp5\n(I1\n(I1\ntp6\ncnumpy\ndtype\np7\n(S'i8'\np8\nI0\nI1\ntp9\nRp10\n(I3\nS'<'\np11\nNNNI-1\nI-1\nI0\ntp12\nbI00\nS'\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00'\np13\ntp14\nb."
>>> p.dump(y)
>>> f.getvalue()
"cnumpy.core.multiarray\n_reconstruct\np0\n(cnumpy\nndarray\np1\n(I0\ntp2\nS'b'\np3\ntp4\nRp5\n(I1\n(I1\ntp6\ncnumpy\ndtype\np7\n(S'i8'\np8\nI0\nI1\ntp9\nRp10\n(I3\nS'<'\np11\nNNNI-1\nI-1\nI0\ntp12\nbI00\nS'\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00'\np13\ntp14\nb.(g5\ntp15\n."
>>> dill.loads(_)
array([1])
>>> 
```

…so I think that logic needs to be pulled into the handling of `numpy` arrays in the `dumps` function.

Can you confirm the above workaround for your own case?

**Comment by mmckerns** [comment]
The above essentially uses `dill`, but avoids a special case `numpy` handler that is hacked into the `Pickler` class as a method.

**Comment by bondarevts** [comment]
I also has the same case and I found the cause of this bug.

The reason is that `dill` function for `numpy.array` objects saving [uses](https://github.com/uqfoundation/dill/blob/5e664956587ff4debfa84a15a9d6a868eee3a1ac/dill/dill.py#L241-L247) nonlocal variable `pik` instead of argument `pickle`. In this case this means that for the second object all data [will be stored in the `Pickler` from the previous object](https://github.com/uqfoundation/dill/blob/5e664956587ff4debfa84a15a9d6a868eee3a1ac/dill/dill.py#L245). 

This happens because function `save_numpy_array` will be created only for the first object, so `pik` closure object will be the same for the second call of function `dumps`.

**Comment by mmckerns** [comment]
fixed in 331ca0fdd645feb3389d30835d0841eabf252f39
